flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -169,6 +172,17 @@ public Builder writeParallelism(int newWriteParallelism) {
       return this;
     }
 
+    /**
+     * Configuring the equality field columns for iceberg table that accept CDC or UPSERT events.
+     *
+     * @param columns defines the iceberg table's key.
+     * @return {@link Builder} to connect the iceberg table.
+     */
+    public Builder equalityFieldColumns(List<String> columns) {
```

Comment:
Do you think that we should consider adding primary key columns to the spec?

Reply:
In the next PR https://github.com/openinx/incubator-iceberg/commit/a863c66eb3d72dd975ea64c75ed2ac35984c17fe,  The flink table SQL's primary key  will act as the equality field columns.   The semantic of iceberg equality columns is almost the same as primary key,  one difference I can think of is:   the uniqueness of key are not enforced. In this [discussion](https://github.com/apache/iceberg/pull/1663#discussion_r528278694),  we don't guarantee the uniqueness when writing a key which has been also wrote in the previous committed txn, that means if : 

```java
Txn-1:  INSERT key1,  txn commit; 
Txn-2:  INSERT key1,  txn commit;
```

Then the table will have two records with the same key. 

If people really need iceberg to maintain the key's uniqueness, then they will need to transform all the `INSERT` to `UPSERT`, which means `DELETE` firstly and then `INSERT` the new values.   

It will introduce another issues:  Each `INSERT` will be regarded as an `UPSERT`,  so it write a `DELETE` and a `INSERT`.  Finally the size of delete files will be almost same as the size of data files.    The process of merging on read will be quite inefficient   because there are too many useless `DELETE` to JOIN.  

The direct way is using bloom filter to reduce the useless `DELETE`, say we will generate bloom filter binary for each committed data file.  When bootstrap the flink/spark job we will need to prefetch all the bloom filter binary from parquet/avro data files's metadata. Before writing a equality delete, we will check the bloom filter, and if the bloom filter indicate that all the committed data files are not containing the given key, then we could skip to append that equality-delete. That would reduce lots of useless `DELETE` in delete files. Of course, the bloom filter will have 'false positive' issue, but that probability is less than 1%, that means we may append
small amout of deletes whose keys don't exist in the current table.  In my view, that should be OK.

In summary, I think it's reasonable to regard those equality fields as primary key in iceberg table, people could choose to use `UNIQUENESS ENFORCED` or `UNIQUENESS NOT-ENFORCED`, in this way they could trade off between strong semantic and performance.

Reply:
For the bloom filter idea, @wangmiao1981 has been working on a proposal for secondary indexes. I think that could be used for the check you're suggesting here.

> people could choose to use UNIQUENESS ENFORCED or UNIQUENESS NOT-ENFORCED, in this way they could trade off between strong semantic and performance.

Are you saying that if uniqueness is enforced, each insert becomes an upsert. But if uniqueness is not enforced, then the sink would assume that whatever is emitting records will correctly delete before inserting? That sounds reasonable to me.

> Finally the size of delete files will be almost same as the size of data files. The process of merging on read will be quite inefficient because there are too many useless DELETE to JOIN.

I think that even if uniqueness is not enforced, tables will quickly require compaction to rewrite the equality deletes. I think we should spend some time making sure that we have good ways to maintain tables and compact equality deletes into position deletes, and position deletes into data files.

Reply:
> Are you saying that if uniqueness is enforced, each insert becomes an upsert. But if uniqueness is not enforced, then the sink would assume that whatever is emitting records will correctly delete before inserting?

Yes.  If someone are exporting relational database's change log events to apache iceberg table and they could guarantee the exactly-once semantics (For example,  the [flink-cdc-connector](https://github.com/ververica/flink-cdc-connectors) could guarantee that), then the uniqueness is always correct when we just write the INSERT/DELETE/UPDATE_BEFORE/UPDATE_AFTER to iceberg.  While in some other cases,  for example flink aggregate job to refresh the metrics count value,  we will write the same key several times without deleting first, then we should regard all the INSERT as UPSERT. 

> even if uniqueness is not enforced, tables will quickly require compaction to rewrite the equality deletes.

That was planned in the second phase, include: 

1.  Use bloom filter to reduce lots of useless deletes; 
2. Minor compaction to convert parts of equality deletes to pos-deletes
3. Major compaction to eliminate all the deletes. 
4. Make the whole read path & write path more stable. For example,  cache policy reduce duplicated delete files loading when merging on read in the same tasks;  Spill to disk if the `insertedRowMap` is exceeding the task's memory threshold, etc.  I will evaluate the read & write & compaction paths in a large dataset, making this to be a stable solution for production. 

It's good to have a document to collect all those things for reviewing.

Reply:
I’d vote for not ensuring uniqueness as it is really hard at scale. If we are to ensure this at write, we have to join the incoming data with the target table making it really expensive. Doing this at read would require sorting the data not only by the sort key but also by the sequence number.